edk2toollib/windows/policy/firmware_policy.py

```diff
@@ -0,0 +1,641 @@
+        Value = PolicyValue.FromBytes(vt, OffsetToValue)
+        return cls(RootKey=RootKey, OffsetToSubKeyName=OffsetToSubKeyName,
+                   OffsetToValueName=OffsetToValueName, OffsetToValue=OffsetToValue,
+                   SubKeyName=SubKeyName, ValueName=ValueName, Value=Value)
+
+    def Print(self, prefix: str = ''):
+        print('%sRule' % (prefix,))
+        print('%s  RootKey:  %x' % (prefix, self.RootKey))
+        print('%s  SubKeyNameOffset:  %x' % (prefix, self.OffsetToSubKeyName))
+        print('%s  ValueNameOffset:  %x' % (prefix, self.OffsetToValueName))
+        print('%s  ValueOffset:  %x' % (prefix, self.OffsetToValue))
+        self.SubKeyName.Print(prefix=prefix + '  SubKeyName ')
+        self.ValueName.Print(prefix=prefix + '  ValueName ')
+        self.Value.Print(prefix=prefix + '  ')
+
+    def Serialize(self, ruleOut: bytearray, valueOut: bytearray, offsetInVT: int):
+        self.OffsetToSubKeyName = offsetInVT
+        localArray = bytearray()
+        self.SubKeyName.Serialize(valueOut=localArray)
+        valueOut += localArray
+
+        self.OffsetToValueName = self.OffsetToSubKeyName + len(localArray)
+        localArray = bytearray()
+        self.ValueName.Serialize(valueOut=localArray)
+        valueOut += localArray
+
+        self.OffsetToValue = self.OffsetToValueName + len(localArray)
+        localArray = bytearray()
+        self.Value.Serialize(valueOut=localArray)
+        valueOut += localArray
+
+        ruleOut += struct.pack(self.StructFormat, self.RootKey, self.OffsetToSubKeyName,
+                               self.OffsetToValueName, self.OffsetToValue)
+
+
+class PolicyValueType():
+    """ Class for storing, serializing, deserializing, & printing PolicyValue Types"""
+    StructFormat = '<H'
+    StructSize = struct.calcsize(StructFormat)
+
+    POLICY_VALUE_TYPE_STRING = 0
+    POLICY_VALUE_TYPE_BOOL = 1
+    POLICY_VALUE_TYPE_DWORD = 2
+    POLICY_VALUE_TYPE_DWORD_RANGE = 3
+    POLICY_VALUE_TYPE_DWORD_CHOICE = 4
+    POLICY_VALUE_TYPE_QWORD = 5
+    POLICY_VALUE_TYPE_QWORD_RANGE = 6
+    POLICY_VALUE_TYPE_QWORD_CHOICE = 7
+    POLICY_VALUE_TYPE_OPTION = 8
+    POLICY_VALUE_TYPE_MULTI_STRING = 9
+    POLICY_VALUE_TYPE_BINARY = 10
+
+    SupportedValueTypes = {POLICY_VALUE_TYPE_DWORD,
+                           POLICY_VALUE_TYPE_QWORD,
+                           POLICY_VALUE_TYPE_STRING}
+
+    def __init__(self, Type):
+        if Type not in self.SupportedValueTypes:
+            ErrorMessage = ('Unsupported ValueType: %x' % Type)
+            print(ErrorMessage)
+            if STRICT:
+                raise Exception(ErrorMessage)
+
+        self.vt = Type
+
+    """if offset is not specified, stream fs position is at beginning PolicyValueType struct"""
+    @classmethod
+    def FromFileStream(cls, fs: BinaryIO, fsOffset: int = None):
+        if fsOffset:
+            fs.seek(fsOffset)
+        valueType = struct.unpack('<H', fs.read(2))[0]
```

Comment:
Can this leverage the FromBytes() method once the file has been read?

Reply:
Good question, you've helped point out where I could make this more understandable.

This method cannot do as you ask because `fsOffset` is the offset from the beginning of the file, where `baOffset` in `FromBytes()` is the offset from the beginning of the ValueTable.

The file can be deserialized sequentially until it encounters the "ValueTable", which requires random access to process.  Streams were originally used, and when the ValueTable was encountered, `seek() `was invoked numerous times to perform the random accesss.
  
This commit 10 incarnation deserializes using streams until it encounters the ValueTable, then to avoid the seeks(), the ValueTable is consumed in 1 read() into Bytes, and further parsing has been switched to use FromBytes().

What I'm inferring is that I might consider reading the whole file into Bytes to begin with, and do all of the parsing on that, instead of stream half the type, bytes half the time?